control-plane/agents/core/src/core/specs.rs

```diff
@@ -55,4 +60,77 @@ impl ResourceSpecsLocked {
             tokio::time::delay_for(period).await;
         }
     }
+
+    /// Completes a volume update operation by trying to update the spec in the persistent store.
+    /// If the persistent store operation fails then the spec is marked accordingly and the dirty
+    /// spec reconciler will attempt to update the store when the store is back online.
+    pub(crate) async fn spec_complete_op<R, O, S: SpecTransaction<O> + StorableObject>(
+        registry: &Registry,
+        result: Result<R, SvcError>,
+        spec: Arc<Mutex<S>>,
+        mut spec_clone: S,
+    ) -> Result<R, SvcError> {
+        match result {
+            Ok(val) => {
+                spec_clone.commit_op();
+                let stored = registry.store_obj(&spec_clone).await;
+                let mut spec = spec.lock().await;
+                match stored {
+                    Ok(_) => {
+                        spec.commit_op();
+                        Ok(val)
+                    }
+                    Err(error) => {
+                        spec.set_op_result(true);
+                        Err(error)
+                    }
+                }
+            }
+            Err(error) => {
+                spec_clone.clear_op();
+                let stored = registry.store_obj(&spec_clone).await;
+                let mut spec = spec.lock().await;
+                match stored {
+                    Ok(_) => {
+                        spec.clear_op();
+                        Err(error)
+                    }
+                    Err(error) => {
+                        spec.set_op_result(false);
+                        Err(error)
+                    }
+                }
+            }
+        }
+    }
+
+    /// Validates the outcome of a update operation .
+    /// In case of an error, it undoes the changes to the spec.
+    /// If the persistent store is unavailable the spec is marked as dirty and the dirty
+    /// spec reconciler will attempt to update the store when the store is back online.
+    pub(crate) async fn spec_step_op<R, O, S: SpecTransaction<O> + StorableObject>(
```

Comment:
I'm not sure what "step" refers to here. Could we call it something like "validate_op"?

Reply:
I guess I was looking at an operation as having multiple "steps". If each intermediate step was Ok(_) then we do nothing and move on to the next one; if it fails we fail the operation.